### PR TITLE
fix Andrei's code to keep the existing DDOC_BLANKLINEs

### PR DESCRIPTION
**Pull request by adamdruppe**
https://github.com/D-Programming-Language/dmd/pull/5315

### CODE REVIEW

**Comment by adamdruppe** [comment]
The main conversation is in Andrei's PR and the base code here is his. I just applied his diff to my code and did a few quick edits to fix some bugs without waiting on him.

I probably should have actually forked his fork and opened a PR against that, then if he accepted it, his pull request would be automatically updated and could be brought right in!

I really think btw that we all ought to start doing that. If you see a PR that isn't quite right... file a PR against the PR's fork/branch. You don't have to limit yourself to just commenting and hoping the original author gets around to it eventually!

File: src/doc.d [hunk 1]
```diff
@@ -2378,6 +2381,18 @@ extern (C++) void highlightText(Scope* sc, Dsymbols* a, OutBuffer* buf, size_t o
             leadingBlank = 0;
             if (sc._module.isDocFile || inCode)
                 break;
+
+
+            if(blankLineRun) {
```
Comment (andralex) [inline src/doc.d]: please edit for style - space after if, brace on own line

File: src/doc.d [hunk 2]
```diff
@@ -2378,6 +2381,18 @@ extern (C++) void highlightText(Scope* sc, Dsymbols* a, OutBuffer* buf, size_t o
...
+                // We got to a non-whitespace in a run of blank lines, time
+                // to insert the paragraph break to group two or more newlines
+                // into a paragraph separator
+                immutable ps = "$(DDOC_PARAGRAPH_SEPARATOR)";
+                i = buf.insert(i, ps.ptr, ps.length) - 1;
+            }
+
+            blankLineRun = 0;
```
Comment (andralex) [inline src/doc.d]: For "efficiency" - heh - move this inside the if.

**Comment by andralex** [comment]
Generally: like. Let's move forward with this. Thanks @adamdruppe!

**Comment by adamdruppe** [comment]
The PR got quite a bit more complex as I got into the nitty-gritty. This is my test file:

```
/++
	This is the summary section.

	This is not a new paragraph. Rather, it is the description section.

	This IS a new paragraph.
	---
	This code sample should not be a paragraph.
	---
	This is a paragraph too.

	&amp; so is this

	`code` can start a paragraph.

	> and so can greater-than symbols




	and this too. There should be only one, despite the run of lines.

	<h1>but this is no p</h1>

	Params:
		args = cmd line arguments

			and more explanation in a paragraph

	Macros:
	DDOC_PARAGRAPH_SEPARATOR=<p>
+/
void main(string[] args) {
}
```

I also tried it with a .dd file:

```
Ddoc
	There is no summary section here, so this is a paragraph.

	No summary, no description. This is a paragraph too.

	This IS a new paragraph.
	---
	This code sample should not be a paragraph.
	---
	This is a paragraph too.

	&amp; so is this

	`code` can start a paragraph.




	and this too. There should be only one, despite the run of lines.

	<h1>but this is no p</h1>

	Macros:
	DDOC_PARAGRAPH_SEPARATOR=<p>
	DDOC_BLANKLINE=
	_= see, in the old code, .dd files had no BLANKLINEs. now they do,
	   so if you don't want them, gotta suppress them yourself.
```

Take special care of the final note there: this changes the default behavior of .dd files!

**Comment by adamdruppe** [comment]
Well, the BLANKLINE thing in `.dd` files failed the bloody tester after all. I guess I'll just leave that as-is. Ugh.

So the other issue with this whole thing is what about:

```
Behold a list:

$(OL
   $(LI what)
)
```

It'll put a paragraph separator before that `OL`.... but is a list supposed to be following a paragraph separator? That's, you guessed it, invalid html. The browser will treat it as an empty paragraph followed by a list, which may mess up margins when rendered - an empty paragraph still takes up space.

We're getting back to the complex rules I was trying to lay out in the other thread, I don't think ddoc can be correct in all cases :(

**Comment by adamdruppe** [comment]
Whelp, I didn't want to take ownership of this because I suspected it would take all day to cover all the edge cases... but I started testing the edge cases and guess what? I was right again! It did take all day :(

But, this code seems to cover everything... and I put in an optional macro for closing tags too because it was actually pretty easy to do once I got it opening them consistently in the right places - and importantly, not opening them in the right places too.

Read the comments in the source code to follow my reasoning. Some of it might be controversial, so the comments there will help future maintainers know WTF I was thinking when I wrote this crap.

The paragraph followed by list example is one where the simple separator idea breaks down. To handle that, I tracked open parenthesis as a proxy of macro activity. If a paragraph begins on a multi-line macro, similarly to the logic of if it begins on a HTML tag, it does not output the separator automatically. It assumes you, the author, wrapped this multi-line block in a macro for a reason.

Authors will need to understand this because a short block macro that happens to be on one line WILL be considered a new paragraph. Sadly, I can't think of any way around this without introducing more serious bugs.

```
     Check this out:

     $(LINK2 something I intended to be inside a paragraph)

    Is indistinguishable from

    $(TIP this shouldn't be in a paragraph but also fit on one line)
```

So I had to make some decision there. But if authors format blocks as

```
    $(TIP
          This is a block!
    )
```

It will avoid the automatic `<p>` tag surrounding the tip and we will be happy.

Inline macros will not terminate a paragraph, but unbalanced parenthesis may. If you don't want to think about the close tags, just don't define it, but I'm pretty sure this is doing the right thing now, including in those unbalanced cases.



I also put the DDOC_BLANKLINE back to the way it used to be because it apparently works well enough for those .dd files now and I don't want fight with the tester. I dread adding these new cases enough as it is.

This is my test file, hopefully I exercised the cases:

```
/++
	This is the summary section.

	This is not a new paragraph. Rather, it is the description section.

	This IS a new paragraph.
	---
	This code sample should not be a paragraph.
	---
	This is a paragraph too.

	---
	this code section is not a new paragraph.
	---


	---
	nor is this one
	---

	&amp; but this is a new paragraph

	`code` can start a paragraph.

	> and so can greater-than symbols




	and this too. There should be only one, despite the run of lines.

	<h1>but this is no p</h1>

	Params:
		args = cmd line arguments

			and more explanation in a paragraph

	List_test:

	$(OL
		$(LI what happens here?

		or here?)
	)

	$(B this is $(I cool)) though because it is an inline macro.

	---
		What about some real code?
		/// $(this rox)
		void someCode() {}
	---

	Macros:
	DDOC_PARAGRAPH_SEPARATOR=<p>
	DDOC_PARAGRAPH_CLOSER=</p>
+/
void main(string[] args) {
}
```

And the .dd test is the same, like above again. blargh this took my entire day and i need to get to bed now. got day job work tomorrow to catch up on.

File: src/doc.d [hunk 3]
```diff
@@ -2116,15 +2118,40 @@ extern (C++) void highlightText(Scope* sc, Dsymbols* a, OutBuffer* buf, size_t o
...
+        {
+            immutable ps = "$(DDOC_PARAGRAPH_SEPARATOR)";
+            i = buf.insert(i, ps.ptr, ps.length);
+            paragraphOpen = true;
+            parensOpenInParagraph = 0;
+        }
+
+        void insertParagraphCloser(size_t where)
+        {
+                if (paragraphOpen)
```
Comment (yebblies) [inline src/doc.d]: Indented too far.

**Comment by yebblies** [comment]
> Read the comments in the source code to follow my reasoning. Some of it might be controversial, so the comments there will help future maintainers know WTF I was thinking when I wrote this crap.

DMD needs more comments like this.

**Comment by andralex** [comment]
As discussed with @adamdruppe, this is too much. We have DDOC_BLANKLINE and should not introduce two more, just one. So please remove DDOC_PARAGRAPH_CLOSER.

**Comment by adamdruppe** [comment]
Superseded by https://github.com/D-Programming-Language/dmd/pull/5319

**Pull request closed without merging**